### DISCUSSION TITLE
How the stm32Xxxx_af.csv files are generated?

### DISCUSSION THREAD

**Discussion by jgpeiro** [micropython/micropython#11508] [STM32 / Pyboard]
I have a board based on stm32g474 (lqfp 100) that is not compatible with other supported CPUs like STM32G473 (lqfp 64).
How can I create the required file? I dont see any python script in the repo. Is this a manual process?
I know that there is a table in the datasheet, but this info is also available in the HAL? (maybe in some device pack files?)
The HAL file stm32Xxx_hal_gpio_ex.h contains lot of AF definitions, but... how they are linked to the pins?

Thanks

**Comment by jimmo** [comment]
Originally I think many of the original ones were generated with a (fairly manual) script that parsed the datasheet PDFs, but I'm not sure that still works.

But yes, the source of information is the table in the ST datasheets.

**Comment by jgpeiro** [comment]
I checked the G474 file and most of pins where compatible. The only difference was the HRTIM that is not present on my G473, but I dont care about this, so I can use this file.

Thanks for answer

**Comment by agesome** [comment]
I got most of the way by finding GPIO-STM32xxx_gpio_v1_0_Modes.xml in STM32Cube install directory and processing it with https://gist.github.com/agesome/cb86b3f110c38f4c03bc658d83dd911c
Filled top row by hand, can probably be sourced from the xml too.

**Comment by andrewleech** [comment]
Also see [stm32/af_csv_gen: Add tool to generate af csv files from STM32_open_pin_data. #12272](https://github.com/micropython/micropython/pull/12272)
